### PR TITLE
Parser: Parse surrogate pairs and es6 unicode code point escapes only when `u` flag is set

### PR DESCRIPTION
**Pull request by Golmote**
I had to mess up with the BNF grammar more than I expected, but it appears to work as intended.

### CODE REVIEW

File: src/parser/regexp.bnf [hunk 1]
```diff
@@ -35,96 +37,112 @@ GROUP_NAME              ([\w$]|\\'u'[0-9a-fA-F]{4}|\\'u{'[0-9a-fA-F]{1,}'}')+
...
+'.'                                 return 'ANY'
 
-{CHAR}                   return 'CHAR'
+'/'                                 return 'SLASH'
 
-'[^'                     { this.pushState('class'); return 'NEG_CLASS' }
-'['                      { this.pushState('class'); return 'L_BRACKET' }
+{CHAR}                              return 'CHAR'
+
+'[^'                                { this.pushState(['u','xu'].includes(this.getCurrentState()) ? 'u_class' : 'class'); return 'NEG_CLASS' }
```
Comment (DmitrySoshnikov) [inline src/parser/regexp.bnf]: Let's not allocate an array each time we entry a class. We can do simple `if` check, or allocate it once at the top as a constant, and use here. You can also create a helper function, and reuse it below as well.

**Comment by DmitrySoshnikov** [comment]
CC @mathiasbynens

File: src/parser/regexp.bnf (hunk 1 above)
Comment (Golmote) [reply src/parser/regexp.bnf]: Ok, changed it.

File: src/parser/regexp.bnf [hunk 2]
```diff
@@ -595,6 +630,9 @@ SourceCharacter
   | ESC_CHAR
     { $$ = Char($1.slice(1), 'simple', @$); $$.escaped = true; }
 
+  | U_CODE_SURROGATE
+    { $$ = Char($1, 'unicode', @$); $$.surrogate_pair = true; }
```
Comment (DmitrySoshnikov) [inline src/parser/regexp.bnf]: let's use JS "camelCase" notation: `isSurrogatePair`

Also, need to update main REAMDE for the `Char` specification for this new field

Comment (Golmote) [reply src/parser/regexp.bnf]: Ok, PR updated!

**Comment by DmitrySoshnikov** [comment]
@Golmote, great job, as always; thanks!

**Pull request merged**